src/bin/sccache-dist/main.rs

```diff
@@ -223,55 +439,93 @@ impl SchedulerIncoming for Scheduler {
     type Error = Error;
     fn handle_alloc_job(&self, requester: &SchedulerOutgoing, tc: Toolchain) -> Result<AllocJobResult> {
         // TODO: prune old servers
-        let server_id = {
-            let servers = self.servers.lock().unwrap();
+        let (job_id, server_id, auth) = {
+            let mut servers = self.servers.lock().unwrap();
             let mut best = None;
             let mut best_load: f64 = MAX_PER_CORE_LOAD;
-            for (id, details) in servers.iter() {
+            let num_servers = servers.len();
+            for (&server_id, details) in servers.iter_mut() {
                 let load = details.jobs_assigned as f64 / details.num_cpus as f64;
                 if load < best_load {
-                    best = Some(id);
+                    best = Some((server_id, details));
                     best_load = load;
                     if load == 0f64 {
                         break
                     }
                 }
             }
-            if let Some(id) = best {
-                *id
+            if let Some((server_id, server_details)) = best {
+                let job_count = self.job_count.fetch_add(1, Ordering::SeqCst) as u64;
+                let job_id = JobId(job_count);
+                server_details.jobs_assigned += 1;
+
+                info!("Job {} created and assigned to server {:?}", job_id, server_id);
+                let mut jobs = self.jobs.lock().unwrap();
+                assert!(jobs.insert(job_id, JobDetail { server_id, state: JobState::Pending }).is_none());
+                let auth = (server_details.generate_job_auth)(job_id);
+                (job_id, server_id, auth)
             } else {
-                let msg = format!("Insufficient capacity: {} available servers", servers.len());
+                let msg = format!("Insufficient capacity across {} available servers", num_servers);
```

Comment:
This seems unfortunate. If a client tries to request a job that we don't have capacity for, it fails? I'd like to chat about this next time we talk.